test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URLDecoderTest.scala

```diff
@@ -0,0 +1,97 @@
+package org.scalajs.testsuite.javalib.net
+
+import org.junit.Test
+import org.junit.Assert._
+
+import java.net.URLDecoder
+import java.io.UnsupportedEncodingException
+
+class URLDecoderTest {
+
+  private final val utf8 = "utf-8"
+  private final val ReplacementChar = '\uFFFD'
+
+  @Test
+  def decodeTest(): Unit = {
+    @inline
+    def test(encoded: String, expected: String, enc: String = utf8): Unit =
+      assertEquals(URLDecoder.decode(encoded, enc), expected)
+
+    @inline
+    def illegalArgument(encoded: String, enc: String = utf8): Unit = {
+      val thrown =
+        try {
+          URLDecoder.decode(encoded, enc)
+          false
+        } catch {
+          case _: IllegalArgumentException => true
+        }
+
+      assertTrue(thrown)
+    }
+
+    @inline
+    def illegalArgumentOrReplacement(encoded: String, enc: String = utf8): Unit = {
+      val thrown =
+        try {
+          val res = URLDecoder.decode(encoded, enc)
+
+          /* it is valid to return the Unicode replacement character (U+FFFD)
+            when encountering an invalid codepoint */
+          res.contains(ReplacementChar)
+        } catch {
+          case _: IllegalArgumentException => true
+        }
+
+      assertTrue(thrown)
+    }
+
+    @inline
+    def unsupportedEncoding(encoded: String, enc: String = utf8): Unit = {
+      val thrown =
+        try {
+          URLDecoder.decode(encoded, enc)
+          false
+        } catch {
+          case _: UnsupportedEncodingException => true
+        }
+
+      assertTrue(thrown)
+    }
+
+    // empty string
+    test("", "")
+
+    // '+' -> ' '
+    test("a+b+c", "a b c")
+
+    // single byte codepoint
+    test("a%20b%20c", "a b c")
+
+    // multi byte codepoint
+    test("a%c3%9fc", "aßc")
+
+    // consecutive characters
+    test("a%20%20c", "a  c")
+
+    // illegal codepoints
+    illegalArgument("a%b%c")
+    illegalArgument("%-1")
```

Comment:
I fixed this, but it seems that the oracle-jdk doesn't follow the spec. 

The spec says: There are two possible ways in which this decoder could deal with illegal strings. It could either leave illegal characters alone or it could throw an IllegalArgumentException. Which approach the decoder takes is left to the implementation

Whereas, what the oracle-jdk does is that it replaces invalid characters with U+FFFD.

From that, I don't know how the tests should be written. Should we base the tests on the specs and skip the ones with illegal characters when run on the JVM. Or should we check for the current behavior URLDecoder in the oracle-jdk (as I did for now).